### ISSUE TITLE
Using behavior class to compare calculators under different growth assumptions/reforms

### ISSUE THREAD

**Issue by derrickchoe** [open-source-economics/Tax-Calculator#1981]
@MattHJensen @martinholmer @andersonfrailey @hdoupe @codykallen 
Hello all,

I'm currently working on a project that involves modelling the effects of additional real growth on charitable giving. I'm making a few modifications to one of my local branches to allow for an income effect on charity in behavior.py, and I am using the code for the income elasticity of taxable income as a guide. 
```
                if calc2.behavior('BE_subinc_wrt_earnings'):
                    # proportional change in after-tax income
                    with np.errstate(invalid='ignore'):
                        pch = np.where(calc1.array('aftertax_income') > 0.,
                                       (calc2.array('aftertax_income') /
                                        calc1.array('aftertax_income')) - 1.,
                                       0.)
                    inc = (calc2.behavior('BE_inc') *
                           pch * calc1.array('e00200'))
                else:
                    # dollar change in after-tax income
                    # Note: combined is f.unit's income+payroll tax liability
                    dch = calc1.array('combined') - calc2.array('combined')
                    inc = calc2.behavior('BE_inc') * dch
```

In the second to last line above, you'll find that the dollar change in after-tax income is simplified as the difference between one's total tax liabilities. However, in cases where one calculator has additional income growth assumptions, this difference should be:
```
calc2.array('aftertax_income') - calc1.array('aftertax_income')
```
Is there any reason that the difference between combined tax liabilities is preferred?

I'm wondering if the behavior class is meant to be used only with calculators under identical growth conditions. If so, I'm happy to model the income effect on charity using different methods. I'm currently using the `behavior.response` function to model the way tax filers respond to both new tax law and income growth. If these two effects are best modeled separately, please let me know! Thanks as always for your advice.

**Comment by martinholmer** [comment]
@derrickchoe said in issue #1981:
> In the second to last line above [in the `Behavior.response` function], you'll find that the dollar change in after-tax income is simplified as the difference between one's total tax liabilities. However, in cases where one calculator has additional income growth assumptions, this difference should be:
```
calc2.array('aftertax_income') - calc1.array('aftertax_income')
```
Is there any reason that the difference between combined tax liabilities is preferred?

Yes, because using the Behavior response parameters is incompatible with using the growdiff_response parameters to simulate "additional income growth".

@derrickchoe continued:
> I'm wondering if the behavior class is meant to be used only with calculators under identical growth conditions. If so, I'm happy to model the income effect on charity using different methods. I'm currently using the Behavior.response function to model the way tax filers respond to both new tax law and income growth. If these two effects are best modeled separately, please let me know!

Yes, you should conduct the behavioral response analysis of charity contributions in runs that do not use growdiff_response parameters, and use growdiff_response parameters (to specify faster than baseline economic growth) in runs that do not use behavioral response parameters.  Will using this approach allow you to conduct the desired analysis?  If not, please explain in more detail exactly what sort of analysis you want to conduct.  We would need that kind of more detailed information to see how Tax-Calculator would need to be enhanced to support the analysis you want to conduct.

Pull requests #1992 and #1993 improve the documentation and checking for behavior/growdiff_response incompatibilities.  Let us know if you have any other questions.

**Comment by derrickchoe** [comment]
@martinholmer thanks for the clarification! I'm also looping in @MattHJensen @codykallen @andersonfrailey @hdoupe . I didn't previously know that the behavior and growdiff options were incompatible with each other.

Some details about my project:
My scholar and I are using taxcalc to estimate the effects of the TCJA on charitable giving for an upcoming paper. I've made some modifications to a local branch allowing for an income effect on charity (based on the proportional change in one's after-tax income). The first part of our analysis doesn't utilize any growdiff parameters. 

However, we also want to estimate the amount of income growth necessary in order to "make up" for the drop in charity as a result of the tax law. I thought that by inflating a number of parameters (as referenced in #1924), I'd be able to measure the change in charitable giving by means of a behavioral response. 

I assumed that the best way to add this "income effect" on charity was through the behavior subclass; I'm open to suggestions as to how to better model this change. Thanks very much.

**Comment by derrickchoe** [comment]
@martinholmer 

I should also mention that @codykallen had previously suggested increasing charity (in response to income growth) through the MPC capabilities in consumption.py. However, the economic models for charitable contributions that I've seen use a log-linear specification and estimate the income effect as an elasticity. As such, I'm more interested in the proportional change in charity in response to a percentage change in income. This was the primary reason that I made modifications to behavior.py on my machine.

The reason that we're interested in economic growth after the TCJA is because we're responding to the argument that charitable giving will be fine under the new law because people will have higher after-tax incomes, both due to direct changes to tax law and increased economic growth. Is there any way to measure a behavioral change in response to both tax reform and alternate growth conditions?

Thanks for all of your help, and I'm looking forward to hearing your suggestions.

**Comment by martinholmer** [comment]
@derrickchoe said:
> However, we also want to estimate the amount of income growth necessary in order to "make up" for the drop in charity as a result of the [TCJA] tax law.  I thought that by inflating a number of parameters (as referenced in #1924), I'd be able to measure the change in charitable giving by means of a behavioral response.
>
> I assumed that the best way to add this "income effect" on charity was through the behavior subclass; I'm open to suggestions as to how to better model this change. 

The Growfactors mentioned in #1924 and documented [here](http://open-source-economics.github.io/Tax-Calculator/#params-growdiff) are used to inflate the Records variables (a couple of which represent charity contributions) each year.  If you want to make charitable contributions grow faster, use the `growdiff_response` parameters that are part of the economic assumptions parameters.  

But this approach will inflate charity contributions **and** other components of income potentially complicating your analysis.  It seems to me that a simpler and cleaner approach is to inflate the `e19800` and `e20100` amounts by a constant before passing the Records object to the Calculator class constructor.  So, you could do something like this:
```
from taxcalc import *
recs = Records.cps_constructor()  # OR "= Records()" IF USING PUF DATA
scale = 1.05
recs.e1980 *= scale
recs.e20100 *= scale
calc = Calculator(policy=Policy(), records=recs)
calc.advance_to_year(2018)  # TCJA is fully implemented in 2018
calc.calc_all()
# tabulate results
```
Vary the value of scale until you find the target charity contributions in your analysis.

Note that the above code is untested, so it could contain errors.  But the idea makes sense to me.
Feel free to ask more questions.

**Comment by derrickchoe** [comment]
@martinholmer 

I appreciate the quick response! You mention:

> But this approach will inflate charity contributions and other components of income potentially complicating your analysis.

Since I'm trying to model the effects of real GDP growth on charity (through increased after-tax incomes), I don't see this as too much of an issue. My original plan was to increase all growdiff_response parameters except for ACPIU, ASOCSEC, and AUCOMP in order to mimic economic growth; then, using the behavior subclass, I could use the percentage change in tax filers' after-tax incomes to estimate the percentage change in charity. 

So in the interest of modelling growth, increasing incomes through growdiff is actually preferable to only inflating charitable contributions by a set rate. As mentioned above, I'm trying to emulate the log-log model for charitable giving as closely as possible. Otherwise, I also would have been happy to increase charitable giving through the MPC of charity in consumption.py.

Is the approach I detail above feasible (i.e. in this specific case, can growdiff and behavior responses be combined to estimate the charity effect of real growth)? If not, I'm happy to explore alternative solutions within taxcalc. Thanks again, and please let me know if my project idea makes sense.

**Comment by martinholmer** [comment]
@derrickchoe said:
> I'm trying to model the effects of real GDP growth on charity (through increased after-tax incomes) ...
> I'm trying to emulate the log-log model for charitable giving as closely as possible.

What's the "log-log model for charitable giving"?  
Have you (or others) done empirical work to estimate the parameters of that model?

**Comment by codykallen** [comment]
@martinholmer asked:

> What's the "log-log model for charitable giving"?

The model is a regression of the log of charitable giving on the log of the tax price of giving and the log of after-tax income. This allows one to say that a percent change in after-tax income, which may be due to a change in the average tax rate or to a change in pre-tax income, has a specific effect on charitable contributions, and that a change in the tax price of giving (`1 + MTR(charity deductions)`). This is somewhat similar to the idea of the income elasticity and the substitution elasticity in use in the Behavior class for the change in taxable income.

I believe that what @derrickchoe is trying to do is essentially to: (1) estimate the percent change in charitable contributions due to the change in the tax price of giving; (2) estimate the percent change in charitable contributions due to the increase in after-tax income from the tax cut (or hike), but using the percent change in after-tax income rather than the dollar change in after-tax income (such as used in the Consumption class's marginal propensity to consume charitable contributions); (3) estimate the percent change in charitable contributions due to an increase in GDP growth. 

This third effect operates through two mechanisms. The increase in GDP causes an increase in incomes, which should increase charitable contributions. At the same time, it also raises the effective marginal tax rates on people (because of a progressive tax system), which increases the effective marginal subsidies on charitable contributions, which reduces the price of giving to charity, which should increase charitable contributions. 

Although in general I agree that the Behavior and growdiff responses should not be used together, I believe that it may be appropriate to do so in order to properly model these three effects on charitable contributions.

**Comment by derrickchoe** [comment]
@martinholmer 
in response to:

> What's the "log-log model for charitable giving"?
Have you (or others) done empirical work to estimate the parameters of that model?

@codykallen covers the vast majority of what I was trying to get at, with a very good explanation. This log-log relationship between income, price, and charitable giving is detailed in the following papers:

http://web.williams.edu/Economics/wp/BakijaHeimCharity.pdf
http://www.nber.org/papers/w3273.pdf

There are numerous other papers that estimate these price and income elasticities.

To build off of Cody's explanation-- our first goal is to measure the change in charitable contributions as a result of the TCJA. Because of changes to marginal tax rates, the increase in the standard deduction, and changes in after-tax income, a filer's giving is affected both by the proportional change in their price of giving, as well as a proportional change in their income. For this part of the analysis, we won't need to incorporate growdiff responses.

Next, we're looking to see how total giving under current (TCJA) law would change if people's incomes increased through real GDP growth. We can increase incomes using growdiff-- we then want to model the accompanying increase in charity through people's behavioral changes (as Cody details above).

Looking forward to getting your thoughts.

Best,

Derrick

**Comment by martinholmer** [comment]
@codykallen and @derrickchoe,  Thanks for all the useful information about the kind of analysis you want to conduct with Tax-Calculator.  I'm still thinking about the best way to make this possible.  But I have a couple preliminary questions.  How are you handling the fact that there are two types of charitable giving (cash and non-cash)?  Are you combining them in the log-log equation or do you specify different log-log-equation elasticities for cash and non-cash.  Also, are you assuming different elasticities for filing units in different income categories?  If so, can you explain?

**Comment by martinholmer** [comment]
@derrickchoe and @codykallen,  Look at pull request #1995, which has been designed to meet your analysis needs.  I'm planning to merge this pull request at the end of the day.  If you have any suggestions, I can add them today or in a subsequent pull request.

**Comment by martinholmer** [comment]
@codykallen said:
> Just to make sure I understand how this works: the `charitable_giving_response` function takes the original charity quantity, price and income elasticities, original and new incomes and marginal tax rates (calculated with the `mtr` function where variable_str = e19800, I presume), and calculates each filing unit's change in charity. So it makes sense to me that this works with `growdiff_response`.

Yes that correct, but I need to clarify the PR a little bit to clear up some issues in what you say above.

When treating cash and non-cash giving the same, the quantity would be the sum of the two.

Yes, I would also presume that the MTR wrt e19800 would be the same as for the non-cash giving.
But, you have to pass the function, the marginal aftertax rate (not simply the MTR).  That's important.

@derrickchoe

**Comment by codykallen** [comment]
@martinholmer, I think you may have mixed up @derrickchoe and me. This is @derrickchoe's project, so he's the best person to handle your questions.

**Comment by martinholmer** [comment]
Sorry for the confusion about who is making the comments.

**Comment by martinholmer** [comment]
@derrickchoe said:
> Understood-- I'll be sure to be inputting an array of 1 - MTR (the price of giving) when using the function. Other than that mix up, I believe I understand how to implement the function in my project.

But aren't the MTRs wrt charitable contributions all zero or negative?
If so, wouldn't the aftertax price be `(1 + MTR)`?

**Comment by derrickchoe** [comment]
@martinholmer said:
> But aren't the MTRs wrt charitable contributions all zero or negative?
If so, wouldn't the aftertax price be (1 + MTR)?

You're right-- I had the sign mixed up. Edited my comment as well.

**Comment by martinholmer** [comment]
Pull request #1995 resolves issue #1981.

**Comment by martinholmer** [comment]
@derrickchoe,  Note that pull request #1997 is renaming the utility function added to support your charitable giving analysis.  There is no change in function logic, just a change in the function's name.  I've also enhanced the function's documentation.

**Comment by martinholmer** [comment]
@derrickchoe, You might be interested in the new Tax-Calculator Cookbook recipe I added in PR #2002.

**Comment by derrickchoe** [comment]
@martinholmer 

Thanks! I'm currently using the `quantity_response` function in my analysis and it greatly simplifies the steps taken to compare different reforms/growth scenarios. I also made a couple of additions to tax calc-- namely, an option for replacing the charitable deduction with a flat rate credit, as well as dollar floors for the amount of charity deductible/subject to a credit. If these would be helpful to add to the repository, I'm happy to make a pull request once the project is finished.

**Comment by martinholmer** [comment]
@derrickchoe said:
> Thanks! I'm currently using the quantity_response function in my analysis and it greatly simplifies the steps taken to compare different reforms/growth scenarios. I also made a couple of additions to tax calc-- namely, an option for replacing the charitable deduction with a flat rate credit, as well as dollar floors for the amount of charity deductible/subject to a credit. If these would be helpful to add to the repository, I'm happy to make a pull request once the project is finished.

I'm glad Tax-Calculator is helping you conduct your tax analysis.  The reform options you mention sound interesting.  We would be interested in considering those enhancements once your project is complete.
Check back with me (here in issue #1981) before you start preparing the pull request.   Thanks.

**Comment by derrickchoe** [comment]
@martinholmer @MattHJensen 

I hope you're doing well! Thanks again for all of your help with estimating these changes in charitable giving. We published [our paper](http://www.aei.org/publication/charitable-giving-and-the-tax-cuts-and-jobs-act/) last month, and I wanted to circle back to some of the taxcalc modifications I made for the analysis. Would any of the following reforms/changes be useful additions to the repository?

1. I added a "charity credit" reform option: this non-refundable credit reduces a taxpayer's tax liability by a set rate multiplied by their charitable giving. This allowed us to model a 25% credit as outlined by the [CBO](https://www.cbo.gov/sites/default/files/112th-congress-2011-2012/reports/charitablecontributions.pdf). 

2. I added a number of options to set a minimum "floor" on charitable contributions subject to a tax benefit; only contributions in excess of this threshold are eligible for a tax deduction/tax credit. This floor can be a dollar value, or a fraction of one's AGI.

3. Lastly, I made a slight modification to the `quantity_response` function. The code currently floors people's after-tax incomes at $1, so that those with negative incomes are counted as having an income of $1. However, this population was shown to have huge swings in charity due to large income effects. If someone with a negative after-tax income went to having an income of $50 under new law, their change in income was calculated as a 4,900% increase-- yielding a huge increase in charitable giving. I controlled for this by only calculating income effects for those with positive after-tax incomes.

If any of these seem interesting, I'm happy to prepare a pull request. I'm looking forward to getting your thoughts.

**Comment by codykallen** [comment]
@derrickchoe, these sound like good potential improvements. However, I recommend moving this discussion to a new issue to avoid confusion.

**Comment by martinholmer** [comment]
@derrickchoe said:
> We published [our paper](http://www.aei.org/publication/charitable-giving-and-the-tax-cuts-and-jobs-act/) last month, and I wanted to circle back to some of the taxcalc modifications I made for the analysis. 
>
> Would any of the following reforms/changes be useful additions to the repository?
>
> 1. I added a "charity credit" reform option: this non-refundable credit reduces a taxpayer's tax liability by a set rate multiplied by their charitable giving. This allowed us to model a 25% credit as outlined by the [CBO](https://www.cbo.gov/sites/default/files/112th-congress-2011-2012/reports/charitablecontributions.pdf). 
>
> 2. I added a number of options to set a minimum "floor" on charitable contributions subject to a tax benefit; only contributions in excess of this threshold are eligible for a tax deduction/tax credit. This floor can be a dollar value, or a fraction of one's AGI.
>
> 3. Lastly, I made a slight modification to the `quantity_response` function. The code currently floors people's after-tax incomes at $1, so that those with negative incomes are counted as having an income of $1. However, this population was shown to have huge swings in charity due to large income effects. If someone with a negative after-tax income went to having an income of $50 under new law, their change in income was calculated as a 4,900% increase-- yielding a huge increase in charitable giving. I controlled for this by only calculating income effects for those with positive after-tax incomes.
>
> If any of these seem interesting, I'm happy to prepare a pull request.

My initial reaction is that we'd be happy to consider pull requests for your item 1 and for your item 2 --- just be certain to put them in separate pull requests.

I'm less certain about your item 3 because I don't understand exactly what you've done.  I'd suggest you open a new Tax-Calculator **issue** that addresses your item 3.  The new issue should describe the problems you ran into in your analysis and show the `quantity_response` code changes you made on your analysis branch.  We can discuss in that issue whether or not your changes to the `quantity_response` code should be incorporated into Tax-Calculator's master branch.

**Comment by derrickchoe** [comment]
@martinholmer 

Thanks for the quick response! You mention: 

> My initial reaction is that we'd be happy to consider pull requests for your item 1 and for your item 2 --- just be certain to put them in separate pull requests.

Currently, my code adds a `CharityCredit` function that takes multiple current law policy parameters and outputs a final credit amount. As an example, if both a dollar floor and a floor as a fraction of AGI are specified, it takes the smaller of the two. The modifications I made for items 1 and 2 are combined in this function; would you suggest that I break up the code in order to open separate pull requests?

In response to: 

>  I'd suggest you open a new Tax-Calculator issue that addresses your item 3. The new issue should describe the problems you ran into in your analysis and show the quantity_response code changes you made on your analysis branch. 

That sounds good. I'll draft some code that shows the results I get under different `quantity_response` versions. Thanks for your suggestion.

**Comment by martinholmer** [comment]
@derrickchoe said:
> Currently, my code adds a CharityCredit function that takes multiple current law policy parameters and outputs a final credit amount. As an example, if both a dollar floor and a floor as a fraction of AGI are specified, it takes the smaller of the two. The modifications I made for items 1 and 2 are combined in this function; would you suggest that I break up the code in order to open separate pull requests?

OK.  I didn't understand you had them so closely intertwined.  
One pull request with both item 1 and item 2 is just fine.
